doc/man/doveadm-penalty.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-PENALTY 1 "2010-07-12" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-penalty \- Show current penalties
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] " penalty " [" \-a
.IR anvil_socket_path ]
[\fIip\fP[\fB/\fP\fImask\fP]]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
.B doveadm penalty
command can be used to see the current penalties.
(Extend me!/explain it)
.\"------------------------------------------------------------------------
@INCLUDE:global-options@
.\" --- command specific options --- "/.
.PP
Command specific
.IR options :
.TP
.BI \-a\  anvil_socket_path
This option is used to specify an alternative socket.
The option\(aqs argument is either an absolute path to a local UNIX domain
socket, or a hostname and port
.RI ( hostname : port ),
in order to connect a remote host via a TCP socket.
.sp
By default
.BR doveadm (1)
will use the socket
.IR @rundir@/anvil .
The socket may be located in another directory, when the default
.I base_dir
setting was overridden in
.IR @pkgsysconfdir@/dovecot.conf .
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.IR ip [/ mask ]
To reduce/filter the output supply an IP address or a network range in
CIDR notation (ip/mask).
.\"------------------------------------------------------------------------
.SH EXAMPLE
Show current penalties
.sp
.nf
.B doveadm penalty
IP               penalty last_penalty        last_update
192.0.2.222            3 2010\-06\-15 15:19:27 15:19:27
192.0.2.53             3 2010\-06\-15 15:19:34 15:19:34
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.1 2018-02-29  Aki Tuomi <aki.tuomi@dovecot.fi>

	* Submission server support improvements and bug fixes
	  - Lots of bug fixes to submission server
	* API CHANGE: array_idx_modifiable will no longer allocate space
	  - Particularly affects how you should check MODULE_CONTEXT result, or
	    use REQUIRE_MODULE_CONTEXT.

	+ mail_attachment_detection_options setting controls when
	  $HasAttachment and $HasNoAttachment keywords are set for mails.
	+ imap: Support fetching body snippets using FETCH (SNIPPET) or
	  (SNIPPET (LAZY=FUZZY))
	+ fs-compress: Automatically detect whether input is compressed or not.
	  Prefix the compression algorithm with "maybe-" to enable the
	  detection, for example: "compress:maybe-gz:6:..."
	+ Added settings to change dovecot.index* files' optimization behavior.
	  See https://wiki2.dovecot.org/IndexFiles#Settings
	+ Auth cache can now utilize auth workers to do password hash
	  verification by setting auth_cache_verify_password_with_worker=yes.
	+ Added charset_alias plugin. See
	  https://wiki2.dovecot.org/Plugins/CharsetAlias
	+ imap_logout_format and pop3_logout_format settings now support all of
	  the generic variables (e.g. %{rip}, %{session}, etc.)
	+ Added auth_policy_check_before_auth, auth_policy_check_after_auth
	  and auth_policy_report_after_auth settings.
	+ master: Support HAProxy PP2_TYPE_SSL command and set "secured"
	  variable appropriately
	- Invalid UCS4 escape in HTML can cause crashes
	- imap: IMAP COMPRESS -enabled clietn crashes on disconnect
	- lmtp: Fix crash when user is over quota
	- lib-lda: Parsing Return-Path header address fails when it contains
	  CFWS
	- auth: SASL with Exim fails for AUTH commands without an initial
	  response
	- imap: SPECIAL-USE capability isn't automatically added
	- auth: LDAP subqueries do not support standard auth variables in
	  var-expand
	- auth: SHA256-CRYPT and SHA512-CRYPT schemes do not work
	- lib-index: mail_always/never_cache_fields are not used for existing
	  cache files
	- imap: Fetching headers leaks memory if search doesn't find any mails
	- lmtp: ORCPT support in RCPT TO
	- imap-login: Process sometimes ends up in infinite loop
	- sdbox: Rolled back save/copy transaction doesn't delete temp files
	- mail: lock_method=dotlock causes crashes

v2.3.0.1 2018-02-28  Timo Sirainen <tss@iki.fi>

	* CVE-2017-15130: TLS SNI config lookups may lead to excessive
	  memory usage, causing imap-login/pop3-login VSZ limit to be reached
	  and the process restarted. This happens only if Dovecot config has
	  local_name { } or local { } configuration blocks and attacker uses
	  randomly generated SNI servernames.
	* CVE-2017-14461: Parsing invalid email addresses may cause a crash or
	  leak memory contents to attacker. For example, these memory contents
	  might contain parts of an email from another user if the same imap
	  process is reused for multiple users. First discovered by Aleksandar
	  Nikolic of Cisco Talos. Independently also discovered by "flxflndy"
	  via HackerOne.
	* CVE-2017-15132: Aborted SASL authentication leaks memory in login
	  process.
	* Linux: Core dumping is no longer enabled by default via
	  PR_SET_DUMPABLE, because this may allow attackers to bypass
	  chroot/group restrictions. Found by cPanel Security Team. Nowadays
	  core dumps can be safely enabled by using "sysctl -w
	  fs.suid_dumpable=2". If the old behaviour is wanted, it can still be
	  enabled by setting:
	  import_environment=$import_environment PR_SET_DUMPABLE=1
	- imap-login with SSL/TLS connections may end up in infinite loop

v2.3.0 2017-12-22  Timo Sirainen <tss@iki.fi>

	* Various setting changes, see https://wiki2.dovecot.org/Upgrading/2.3
	* Logging rewrite started: Logging is now based on hierarchical events.
	  This makes it possible to do various things, like: 1) giving
	  consistent log prefixes, 2) enabling debug logging with finer
	  granularity, 3) provide logs in more machine readable formats
	  (e.g. json). Everything isn't finished yet, especially a lot of the
	  old logging code still needs to be translated to the new way.
	* Statistics rewrite started: Stats are now based on (log) events.
	  It's possible to gather statistics about any event that is logged.
	  See http://wiki2.dovecot.org/Statistics for details
	* ssl_dh setting replaces the old generated ssl-parameters.dat
	* IMAP: When BINARY FETCH finds a broken mails, send [PARSE] error
	  instead of [UNKNOWNCTE]
	* Linux: core dumping via PR_SET_DUMPABLE is no longer enabled by
	  default due to potential security reasons (found by cPanel Security
	  Team).

	+ Added support for SMTP submission proxy server, which includes
	  support for BURL and CHUNKING extension.
	+ LMTP rewrite. Supports now CHUNKING extension and mixing of
	  local/proxy recipients.
	+ auth: Support libsodium to add support for ARGON2I and ARGON2ID
	  password schemes.
	+ auth: Support BLF-CRYPT password scheme in all platforms
	+ auth: Added LUA scripting support for passdb/userdb.
	  See https://wiki2.dovecot.org/AuthDatabase/Lua
	- Input streams are more reliable now when there are errors or when
	  the maximum buffer size is reached. Previously in some situations
	  this could have caused Dovecot to try to read already freed memory.
	- Output streams weren't previously handling failures when writing a
	  trailer at the end of the stream. This mainly affected encrypt and
	  zlib compress ostreams, which could have silently written truncated
	  files if the last write happened to fail (which shouldn't normally
	  have ever happened).
	- virtual plugin: Fixed panic when fetching mails from virtual
	  mailboxes with IMAP BINARY extension.
	- doveadm-server: Fix potential hangs with SSL connections
	- doveadm proxy: Reading commands' output from v2.2.33+ servers could
	  have caused the output to be corrupted or caused a crash.
	- Many other smaller fixes

v2.2.33.2 2017-10-20  Timo Sirainen <tss@iki.fi>


src/doveadm/doveadm-penalty.c
/* Copyright (c) 2009-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "array.h"
#include "net.h"
#include "istream.h"
#include "hash.h"
#include "strescape.h"
#include "time-util.h"
#include "doveadm.h"
#include "doveadm-print.h"

#include <unistd.h>

struct penalty_line {
	struct ip_addr ip;
	unsigned int penalty;
	time_t last_penalty, last_update;
};

struct penalty_context {
	const char *anvil_path;

	struct ip_addr net_ip;
	unsigned int net_bits;
};

static void penalty_parse_line(const char *line, struct penalty_line *line_r)
{
	const char *const *args = t_strsplit_tabescaped(line);
	const char *ident = args[0];
	const char *penalty_str = args[1];
	const char *last_penalty_str = args[2];
	const char *last_update_str = args[3];

	i_zero(line_r);

	(void)net_addr2ip(ident, &line_r->ip);
	if (str_to_uint(penalty_str, &line_r->penalty) < 0 ||
	    str_to_time(last_penalty_str, &line_r->last_penalty) < 0 ||
	    str_to_time(last_update_str, &line_r->last_update) < 0)
		i_fatal("Read invalid penalty line: %s", line);
}

static void
penalty_print_line(struct penalty_context *ctx,
		   const struct penalty_line *line)
{
	if (ctx->net_bits > 0) {
		if (!net_is_in_network(&line->ip, &ctx->net_ip, ctx->net_bits))
			return;
	}

	doveadm_print(net_ip2addr(&line->ip));
	doveadm_print(dec2str(line->penalty));
	doveadm_print(unixdate2str(line->last_penalty));
	doveadm_print(t_strflocaltime("%H:%M:%S", line->last_update));
}

static void penalty_lookup(struct penalty_context *ctx)
{
#define ANVIL_HANDSHAKE "VERSION\tanvil\t1\t0\n"
#define ANVIL_CMD ANVIL_HANDSHAKE"PENALTY-DUMP\n"
	struct istream *input;
	const char *line;
	int fd;

	fd = doveadm_connect(ctx->anvil_path);
	net_set_nonblock(fd, FALSE);
	if (write(fd, ANVIL_CMD, strlen(ANVIL_CMD)) < 0)
		i_fatal("write(%s) failed: %m", ctx->anvil_path);

	input = i_stream_create_fd_autoclose(&fd, (size_t)-1);
	while ((line = i_stream_read_next_line(input)) != NULL) {
		if (*line == '\0')
			break;
		T_BEGIN {
			struct penalty_line penalty_line;

			penalty_parse_line(line, &penalty_line);
			penalty_print_line(ctx, &penalty_line);
		} T_END;
	}
	if (input->stream_errno != 0) {
		i_fatal("read(%s) failed: %s", ctx->anvil_path,
			i_stream_get_error(input));
	}

	i_stream_destroy(&input);
}

static void cmd_penalty(struct doveadm_cmd_context *cctx)
{
	struct penalty_context ctx;
	const char *netmask;

	i_zero(&ctx);
	if (!doveadm_cmd_param_str(cctx, "socket-path", &(ctx.anvil_path)))
		ctx.anvil_path = t_strconcat(doveadm_settings->base_dir, "/anvil", NULL);

	if (doveadm_cmd_param_str(cctx, "netmask", &netmask)) {
		if (net_parse_range(netmask, &ctx.net_ip, &ctx.net_bits) != 0) {
			doveadm_exit_code = EX_USAGE;
			i_error("Invalid netmask '%s' given", netmask);
			return;
		}
	}

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header_simple("IP");
	doveadm_print_header_simple("penalty");
	doveadm_print_header_simple("last_penalty");
	doveadm_print_header_simple("last_update");

	penalty_lookup(&ctx);
}

struct doveadm_cmd_ver2 doveadm_cmd_penalty_ver2 = {
	.name = "penalty",
	.cmd = cmd_penalty,
	.usage = "[-a <anvil socket path>] [<ip/bits>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('a',"socket-path", CMD_PARAM_STR,0)
DOVEADM_CMD_PARAM('\0',"netmask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};
